### PR TITLE
add Boland DF estimation

### PR DESCRIPTION
**Pull request by mikofski**
- [x] referenced in Google group: https://groups.google.com/g/pvlib-python/c/TsDIiU19AUU/m/Bwi9595PAwAJ
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries to [`docs/sphinx/source/api.rst`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/api.rst) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

`pvlib.irradiance.boland` is another diffuse fraction, DF, estimation method similar to Erbs but uses a single logistic exponential correlation between DF and clearness index, kt, that is continuously differentiable and bounded between zero and one.

### CODE REVIEW

**Comment by mikofski** [comment]
Hi @cwhanse, interested in reviewing the gallery example and added Boland DNI estimation function? Thanks

File: pvlib/irradiance.py
Comment (cwhanse) [inline pvlib/irradiance.py:2195]:
```suggestion
    The Boland model [1]_, [2]_ estimates the diffuse fraction, DF, from global
    horizontal irradiance, GHI, through an empirical relationship between DF
    and the clearness index, :math:`k_t`, the ratio of GHI to extraterrestrial irradiance.
```

File: pvlib/irradiance.py
Comment (cwhanse) [inline pvlib/irradiance.py:2201]: This line isn't needed with the above comment

File: pvlib/irradiance.py
Comment (cwhanse) [inline pvlib/irradiance.py:2206]:
```suggestion
        Global horizontal irradiance. [W/m^2]
```

File: pvlib/irradiance.py
Comment (cwhanse) [inline pvlib/irradiance.py:2205]: Remind me, does `numeric` include `Series`?

File: pvlib/irradiance.py
Comment (cwhanse) [inline pvlib/irradiance.py:2209]:
```suggestion
    datetime_or_doy : int, float, array, DatetimeIndex
```

File: docs/examples/plot_diffuse_fraction.py
Comment (cwhanse) [inline docs/examples/plot_diffuse_fraction.py:15]:
```suggestion
# irradiance (GHI). Irradiance sensors such as pyranometers typically only
```

File: docs/examples/plot_diffuse_fraction.py
Comment (cwhanse) [inline docs/examples/plot_diffuse_fraction.py:18]:
```suggestion
# measure GHI. pvlib provides several functions that can be used
# to separate GHI into the diffuse and direct components. The separate
# components are needed to estimate the total irradiance on a tilted
# surface.
```

File: docs/examples/plot_diffuse_fraction.py
Comment (cwhanse) [inline docs/examples/plot_diffuse_fraction.py:44]: Add a introduction something like 
```
# Methods for separating DHI into diffuse and direct components include: **DISC, DIRINT, Erbs** and **Boland**.
```

File: docs/examples/plot_diffuse_fraction.py
Comment (cwhanse) [inline docs/examples/plot_diffuse_fraction.py:47]: Should this be `:py:func:` I'm not at all clear on usage of these markups.

File: docs/examples/plot_diffuse_fraction.py
Comment (cwhanse) [inline docs/examples/plot_diffuse_fraction.py:41]: Maybe add a comment explaining the `*100` on the pressure (conversion from millibar to Pa, I assume)

File: docs/examples/plot_diffuse_fraction.py
Comment (cwhanse) [inline docs/examples/plot_diffuse_fraction.py:173]: The conclusions should include a statement to the effect that DNI and DHI in the TMY3 file are themselves the output of a model, and so difference between e.g. DISC output and TMY DNI shouldn't be regarded as an error in the DISC model. Without the caveat, casual readers would assume that the four models should reproduce the TMY values, which is not a reasonable expectation. We could refer those interested to the TMY3 model documentation [here](https://www.nrel.gov/docs/fy12osti/54824.pdf)). 

If we want to draw conclusions about the accuracy of the various separation models, we should add a measured, three component data set to pvlib, use it here with the caveat that all four models are fit to other data, so perfect agreement shouldn't be assumed.

File: docs/examples/plot_diffuse_fraction.py
Comment (cwhanse) [inline docs/examples/plot_diffuse_fraction.py:186]: Suggested revision
```
# horizontal component of the extra-terrestrial irradiance. At low
# sun elevation (zenith near 90 degrees), especially near sunset, kt can
# explode because the denominator (extra-terrestrial irradiance) approaches
# zero. In pvlib this behavior is moderated by ``min_cos_zenith`` and ``max_clearness_index``
# which each have reasonable defaults. Even so, near sunset there are still spikes in DNI
# for Jan. 5th & 7th, April 4th, 5th, & 7th, and July 6th & 7th. 

# By contrast, the DISC and DIRINT methods estimate DNI first by means of
# correlations, which include additional variables such as airmass. These methods
# seem to reduce DNI spikes over 1000 [W/m^2].
```

This statement needs to go, "error" isn't synonymous with different from TMY3 DNI:
```
, but still have errors at other times.
```


I think these points are covered in the edited paragraph above:

```
#
# Another difference is that DISC & DIRINT return DNI whereas Erbs & Boland
# calculate the diffuse fraction which is then used to derive DNI from GHI and
# the solar zenith, which exacerbates errors at low sun elevation due to the
# relation: DNI = GHI*(1 - DF)/cos(zenith).
```

**Review by cwhanse — COMMENTED** [review]
Looks pretty good, @mikofski .  Thanks for the addition.

File: pvlib/irradiance.py
Comment (mikofski) [reply pvlib/irradiance.py:2288]: I think this is a good idea, but perhaps out of scope for boland pr? I would prefer to open a new pr: "use ghi index for doy if none given in irradiance decomposition functions" if that's okay?

File: pvlib/irradiance.py
Comment (mikofski) [reply pvlib/irradiance.py:2303]: Debatable, but I think it should be a separate issue/discussion/PR and depending on consensus, applied to all decomposition functions like Erbs, etc. at the same time.

File: pvlib/irradiance.py
Comment (mikofski) [reply pvlib/irradiance.py:2285]: ok, but we should update other decomposition functions to have same convention.

File: pvlib/irradiance.py
Comment (mikofski) [reply pvlib/irradiance.py:2284]: TL;DR: Yes, if the `pd.Series` is an array of a numeric type like float, integer, boolean, or complex, otherwise no. _EG_: a pandas series of strings would _not_ be numeric.

I can't find where "numeric" is defined except in NumPy here:
> There are 5 basic numerical types representing booleans (`bool`), integers (`int`), unsigned integers (`uint`) floating point (`float`) and complex.

and in [pandas `to_numeric()`](https://pandas.pydata.org/docs/reference/api/pandas.to_numeric.html).

The "numeric" data type is used extensively in pvlib and I believe its use is consistent with the NumPy & pandas definition, and I don't believe it is constrained to either scalars or arrays.

File: pvlib/irradiance.py
Comment (mikofski) [reply pvlib/irradiance.py:2288]: I prefer `pandas.DatetimeIndex` to use the intersphinx extension so that it's automatically hyperlinked to pandas documentation. Ditto for `numpy.ndarray`

**Comment by mikofski** [comment]
Hi @cwhanse sorry for the very long delay in responding. I have accepted your changes, added discussions for your suggestions, and responded to your comments. Let me know if you have any additional feedback. Thanks!

**Comment by mikofski** [comment]
BTW: I think the readthedocs failure is unrelated, I'm trying to figure it out - the bifacial gallery examples are failing with this error:
>ImportError: cannot import name 'geos_geometrycollection_from_py' from 'shapely.geometry.collection' (/home/docs/checkouts/readthedocs.org/user_builds/pvlib-python/envs/1179/lib/python3.7/site-packages/shapely/geometry/collection.py)

**Comment by mikofski** [comment]
looks like we do not have reproducible readthedocs builds - see this: https://docs.readthedocs.io/en/stable/guides/reproducible-builds.html

**Comment by mikofski** [comment]
@kanderso-nrel I've figured out why readthedocs is going bonkers. For some reason the version is being read as pvlib-0.9.0a5.dev58+g<commit-SHA> EG: the last commit (9715084):
image

This only started happening after I merged with main 9d9421e, but oddly, that was the last commit that _did_ have the correct version: pvlib-0.9.5.dev16+g9d9421e
image

And readthedocs _did_ build. This has something to do with the pep517/518 changes, `pyproject.toml` and `importlib_metadata` but none of these files are affected as far as I can tell.

When the version is misread as v0.9.0, that causes pvfactors to downgrade, and that causes shapely>v2 to be installed leading to the build failure.

I don't understand how the new build system works, all I can hope is that it all works out when/if this gets merged.

**Comment by mikofski** [comment]
You can view a rendering of the docs here: https://pvlib-python--1179.org.readthedocs.build/en/1179/

I guess the readthedocs failures has something to do with [`setuptools_scm`](https://github.com/pypa/setuptools_scm). I had to delete the `pvlib.egg-info` file in my local working copy to get it to return the correct version:
```python
from importlib_metadata import version
import pvlib
version('pvlib')
# '0.9.5.dev34+g1ae35cd'
```

Before I purged the egg-info file, it was returning 0.9.0!

**Comment by kanderso-nrel** [comment]
@mikofski I think the problem was that RTD does a shallow clone by default and didn't give `setuptools_scm` enough info to determine the correct version string for this PR (which has a long/strange pedigree now?  not sure...)

Anyway to get around this problem we can do "unshallow" clones instead (see [docs](https://docs.readthedocs.io/en/stable/build-customization.html#unshallow-git-clone)).  I hope you don't mind that I went ahead and pushed a fix directly to this branch!

File: pvlib/irradiance.py
Comment (cwhanse) [inline pvlib/irradiance.py:2293]:
```suggestion
        clearness index :math:`k_t`. Equivalent to zenith = 86.273 degrees.
```

**Review by cwhanse — APPROVED** [review]

**Comment by AdamRJensen** [comment]
I'm out back country skiing but I'd like to review on Monday if it isn't merged by then 😎

File: docs/examples/plot_diffuse_fraction.py
Comment (kanderso-nrel) [inline docs/examples/plot_diffuse_fraction.py:136]: Nit: these variable names, and the analogous ones below, don't seem to exactly match their values.

**Review by kanderso-nrel — COMMENTED** [review]
Thanks @mikofski, some comments below :)

File: pvlib/irradiance.py
Comment (wholmgren) [reply pvlib/irradiance.py:2335]: For now, I recommend a kwarg `coefficients` and including these in the Notes section of the docstring. A follow up could add an option to infer the appropriate coefficients. Inference should definitely be in a function `boland_diffuse_fraction`. I'd like to see that function in this PR too, but I won't push it.

File: pvlib/irradiance.py
Comment (adriesse) [inline pvlib/irradiance.py:2343]: I know these check have been copied a few times (along with most of the function) and I told myself a few times I should look at them more closely one day...  If ghi were clipped at zero and zenith constrained as advertised, could there still be any bad values?

File: pvlib/irradiance.py
Comment (mikofski) [reply pvlib/irradiance.py:2345]: TBH, IDK, but all of the functions in this module use this. I'll open a separate issue to respond to this.

File: pvlib/irradiance.py
Comment (mikofski) [reply pvlib/irradiance.py:2335]: I rearranged to formula to match the reference, as well as the coefficients, which had been rounded to 8.6 and 5 (=5.3=8.645*0.613).

File: docs/examples/plot_diffuse_fraction.py
Comment (mikofski) [reply docs/examples/plot_diffuse_fraction.py:2]: moved it to new subdirectory called "irradiance-decomposition"

File: pvlib/tests/test_irradiance.py
Comment (mikofski) [reply pvlib/tests/test_irradiance.py:820]: I copied this from other decomposition functions in the irradiance.py module. I changed boland to use `assert np.allclose()`

File: pvlib/irradiance.py
Comment (mikofski) [reply pvlib/irradiance.py:2343]: see #1685

**Comment by mikofski** [comment]
@kanderso-nrel thanks for tips. I think I've addressed everything. @AdamRJensen looking forward to your comments.

File: pvlib/irradiance.py
Comment (adriesse) [reply pvlib/irradiance.py:2351]: I think if we're going to call it a boland "model", the coefficients are part of the model.  If they are not, we can call it logistic function model and offer the boland "coefficients" to accompany it.

In my parallel universe I call it the boland model and offer a "period"option that can have values of 15 or 60.

File: pvlib/irradiance.py
Comment (adriesse) [reply pvlib/irradiance.py:2361]: In favor of regular dict.

File: pvlib/irradiance.py
Comment (cwhanse) [reply pvlib/irradiance.py:2351]: I'd favor renaming to `logistic` and naming coefficient sets, if there was another author that fit this same equation. I'm not aware of any.

**Review by AdamRJensen — APPROVED** [review]
Great stuff @mikofski!

Only a few minor comments from my side.

File: pvlib/irradiance.py
Comment (AdamRJensen) [reply pvlib/irradiance.py:2280]: Further below where the coefficients are provided, lowercase letters are also used fyi.

Comment (mikofski) [inline pvlib/irradiance.py:2280]:
```suggestion
        \mathit{DF} = \frac{1}{1 + \exp\left(a \left(k_t - b\right)\right)}
```

File: docs/sphinx/source/reference/irradiance/decomposition.rst
Comment (mikofski) [reply docs/sphinx/source/reference/irradiance/decomposition.rst:15]: See discussion https://github.com/pvlib/pvlib-python/discussions/1696

**Comment by mikofski** [comment]
@pvlib/pvlib-maintainer ready to merge

**Pull request merged**
